engine/execution/rpc/engine.go

```diff
@@ -269,18 +269,43 @@ func (h *handler) GetEventsForBlockIDs(
 			return nil, status.Errorf(codes.Internal, "state commitment for block ID %s could not be retrieved", bID)
 		}
 
-		// lookup events
-		blockEvents, err := h.events.ByBlockIDEventType(bID, flow.EventType(eType))
+		// lookup all events for the block
+		blockAllEvents, err := h.events.ByBlockID(bID)
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "failed to get events for block: %v", err)
 		}
 
+		// filter events by type
+		eventType := flow.EventType(eType)
+		blockEvents := make([]flow.Event, 0, len(blockAllEvents))
+		for _, event := range blockAllEvents {
+			if event.Type == eventType {
+				blockEvents = append(blockEvents, event)
+			}
+		}
+
 		result, err := h.eventResult(bID, blockEvents)
 		if err != nil {
 			return nil, err
 		}
 		results[i] = result
 
+		// additional check that when there is no event in the block, double check if the execution
```

Comment:
since this doesn't rely on `results` and `blockEvents`, you could move this first so it fails fast